examples/resource_partitioner/async_customization.cpp

```diff
@@ -0,0 +1,534 @@
+    ->  future<typename util::detail::invoke_deferred_result<
+        F, Future, Ts...>::type>
+    {
+        typedef typename util::detail::invoke_deferred_result<
+                F, Future, Ts...>::type result_type;
+
+        using namespace hpx::util::debug;
+        std::cout << "then_execute : Function     : "
+                  << print_type<F>() << "\n";
+        std::cout << "then_execute : Predecessor  : "
+                  << print_type<Future>() << "\n";
+        std::cout << "then_execute : Future       : "
+                  << print_type<typename traits::future_traits<Future>::result_type>()
+                  << "\n";
+        std::cout << "then_execute : Arguments    : "
+                  << print_type<Ts...>(" | ") << "\n";
+        std::cout << "then_execute : Result       : "
+                  << print_type<result_type>() << "\n";
+
+        // forward the task on to the 'real' underlying executor
+        lcos::local::futures_factory<result_type()> p(
+            executor_,
+            util::deferred_call(std::forward<F>(f),
+                                std::forward<Future>(predecessor),
+                                std::forward<Ts>(ts)...)
+        );
+
+        p.apply(
+            launch::async,
+            threads::thread_priority_default,
+            threads::thread_stacksize_default);
+
+        return p.get_future();
+    }
+
+    // --------------------------------------------------------------------
+    // .then() execute specialized for a when_all dispatch for any future types
+    // future< tuple< is_future<a>::type, is_future<b>::type, ...> >
+    // --------------------------------------------------------------------
+    template <typename F,
+              template <typename> typename  OuterFuture,
+              typename ... InnerFutures,
+              typename ... Ts,
+              typename = typename std::enable_if_t<is_future_of_tuple_of_futures<
+                OuterFuture<util::tuple<InnerFutures...>>>::value>,
+              typename = typename std::enable_if_t<is_tuple_of_futures<
+                util::tuple<InnerFutures...>>::value>
+              >
+    auto
+    then_execute(F && f,
+                 OuterFuture<util::tuple<InnerFutures... > >&& predecessor,
+                 Ts &&... ts)
+    ->  future<typename util::detail::invoke_deferred_result<
+        F, OuterFuture<util::tuple<InnerFutures... >>, Ts...>::type>
+    {
+        typedef typename util::detail::invoke_deferred_result<
+            F, OuterFuture<util::tuple<InnerFutures... >>, Ts...>::type
+                result_type;
+
+        // get the tuple of futures from the predecessor future <tuple of futures>
+        const auto & predecessor_value = future_extract_value().operator()(predecessor);
+
+        // create a tuple of the unwrapped future values
+        auto unwrapped_futures_tuple = util::map_pack(
+            future_extract_value{},
+            predecessor_value
+        );
```

Comment:
If I am not mistaken, this `then_execute` overload has more of a `wait_all` semantic, right?

Reply:
yes, if you look at the comment at the start of the function, you see this is a specialization for when_all semantics